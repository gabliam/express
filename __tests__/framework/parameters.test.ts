import {
  Controller,
  Get,
  RequestParam,
  Request,
  Response,
  QueryParam,
  Post,
  RequestBody,
  ExpressConfig,
  RequestHeaders,
  Cookies,
  Next
} from '../../src/index';
import * as bodyParser from 'body-parser';
import * as cookieParser from 'cookie-parser';
import * as e from 'express';
import { ExpressPluginTest } from '../express-plugin-test';
import * as supertest from 'supertest';
import { Config } from '@gabliam/core';
import * as sinon from 'sinon';

let appTest: ExpressPluginTest;

beforeEach(async () => {
  appTest = new ExpressPluginTest();
});

afterEach(async () => {
  await appTest.destroy();
});

describe('Parameters:', () => {
  test('should bind a method parameter to the url parameter of the web request', async () => {
    @Controller('/')
    class TestController {
      @Get(':id')
      public getTest(
        @RequestParam('id') id: string,
        req: e.Request,
        res: e.Response
      ) {
        return id;
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app).get('/foo').expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the request object', async () => {
    @Controller('/')
    class TestController {
      @Get(':id')
      public getTest(@Request() req: e.Request) {
        return req.params.id;
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app).get('/GET').expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the response object', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(@Response() res: e.Response) {
        return res.send('foo');
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to a query parameter', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(@QueryParam('id') id: string) {
        return id;
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/')
      .query('id=lolilol')
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the request body', async () => {
    @Controller('/')
    class TestController {
      @Post('/')
      public getTest(@RequestBody() reqBody: string) {
        return reqBody;
      }
    }

    @Config()
    class ServerConfig {
      @ExpressConfig()
      serverConfig(app: e.Application) {
        app.use(bodyParser.json());
      }
    }

    appTest.addClass(TestController);
    appTest.addClass(ServerConfig);
    await appTest.build();
    const response = await supertest(appTest.app)
      .post('/')
      .send({ foo: 'bar' })
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the request headers', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(@RequestHeaders('testhead') headers: any) {
        return headers;
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/')
      .set('TestHead', 'fooTestHead')
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to a cookie', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getCookie(
        @Cookies('cookie') cookie: any,
        req: e.Request,
        res: e.Response
      ) {
        if (cookie) {
          res.send(cookie);
        } else {
          res.send(':(');
        }
      }
    }

    @Config()
    class ServerConfig {
      @ExpressConfig()
      serverConfig(app: e.Application) {
        app.use(cookieParser());
        app.use(function(req, res, nextFunc) {
          res.cookie('cookie', 'hey');
          nextFunc();
        });
      }
    }

    appTest.addClass(TestController);
    appTest.addClass(ServerConfig);
    await appTest.build();
    const response = await supertest(appTest.app).get('/').expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the next function', async () => {
    const spy = sinon.spy();

    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(@Next() nextFunc: any) {
        spy();
        return nextFunc();
      }

      @Get('/')
      public getResult() {
        return 'foo';
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app).get('/').expect(200);
    expect(spy.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });
});
