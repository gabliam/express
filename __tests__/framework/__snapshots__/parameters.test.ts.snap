// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parameters: should bind a method parameter to a cookie 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "2",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"2-vyGp6PvFo4RvsFtPoIWeCReyIC8\\"",
    "set-cookie": Array [
      "cookie=hey; Path=/",
    ],
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "{}",
}
`;

exports[`Parameters: should bind a method parameter to a query parameter 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "7",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"7-5Uyg9FH6Z639wlnjohqGsan13Gc\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "lolilol",
}
`;

exports[`Parameters: should bind a method parameter to a query parameter and cast to number 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-+eIvcEC3YUszYERL17BmmrIVMVw\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "[\\"number\\",12]",
}
`;

exports[`Parameters: should bind a method parameter to the next function 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-C+7Hteo/D9vJXQ3UfzxbwnXaijM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "foo",
}
`;

exports[`Parameters: should bind a method parameter to the request body 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-pedE0BZFQNM7HX6mFsKPL6l+dUo\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": Object {
      "foo": "bar",
    },
    "headers": Object {
      "content-type": "application/json",
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "POST",
  },
  "status": 200,
  "text": "{\\"foo\\":\\"bar\\"}",
}
`;

exports[`Parameters: should bind a method parameter to the request headers 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "11",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"b-zaZ1r0YCMBDq+OjlsZnp2JpqfDA\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "testhead": "fooTestHead",
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "fooTestHead",
}
`;

exports[`Parameters: should bind a method parameter to the request object 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Parameters: should bind a method parameter to the response object 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-C+7Hteo/D9vJXQ3UfzxbwnXaijM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "foo",
}
`;

exports[`Parameters: should bind a method parameter to the url parameter of the web request 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-C+7Hteo/D9vJXQ3UfzxbwnXaijM\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "foo",
}
`;

exports[`Parameters: should bind a method parameter to the url parameter of the web request and cast to number 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-f7JRWtNSw/8B+r0iQq6BpCFihJo\\"",
    "x-powered-by": "Express",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "user-agent": "node-superagent/3.5.2",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "[\\"number\\",42]",
}
`;
